tests/integration/test_task_pool.py

```diff
@@ -825,3 +819,102 @@ async def test_reload_prereqs(
             ),
             key=lambda d: tuple(d.keys())[0],
         ) == expected_4
+
+
+async def _test_restart_prereqs_sat():
+    # YIELD: the workflow has now started...
+    schd = yield
+
+    # Release tasks 1/a and 1/b
+    schd.pool.release_runahead_tasks()
+    schd.release_queued_tasks()
+    assert list_tasks(schd) == [
+        ('1', 'a', 'running'),
+        ('1', 'b', 'running')
+    ]
+
+    # Mark both as succeeded and spawn 1/c
+    for itask in schd.pool.get_all_tasks():
+        itask.state_reset('succeeded')
+        schd.pool.spawn_on_output(itask, 'succeeded')
+        schd.workflow_db_mgr.put_insert_task_outputs(itask)
+        schd.pool.remove_if_complete(itask)
+    schd.workflow_db_mgr.process_queued_ops()
+    assert list_tasks(schd) == [
+        ('1', 'c', 'waiting')
+    ]
+
+    # YIELD: the workflow has now restarted or reloaded with the new config...
+    schd = yield
+    assert list_tasks(schd) == [
+        ('1', 'c', 'waiting')
+    ]
+
+    # Check resulting dependencies of task z
+    task_c = schd.pool.get_all_tasks()[0]
+    assert sorted(
+        (*key, satisfied)
+        for prereq in task_c.state.prerequisites
+        for key, satisfied in prereq.satisfied.items()
+    ) == [
+        ('1', 'a', 'succeeded', 'satisfied naturally'),
+        ('1', 'b', 'succeeded', 'satisfied naturally')
+    ]
+
+
+@pytest.mark.parametrize('do_restart', [True, False])
+async def test_graph_change_prereq_satisfaction(
+    flow, scheduler, start, do_restart
+):
+    """It should handle graph prerequisites change on reload/restart.
+
+    If the graph is changed to add a dependency which has been previously
+    satisfied, then Cylc should perform a DB check and mark the prerequsite
+    as satisfied accordingly.
+
+    See https://github.com/cylc/cylc-flow/pull/5334
+
+    """
+    conf = {
+        'scheduler': {'allow implicit tasks': 'True'},
+        'scheduling': {
+            'graph': {
+                'R1': '''
+                    a => c
+                    b
+                '''
+            }
+        }
+    }
+    id_ = flow(conf)
+    schd = scheduler(id_, run_mode='simulation', paused_start=False)
+
+    test = _test_restart_prereqs_sat()
+    test.asend(None)
+
+    if do_restart:
+        async with start(schd):
+            # start the workflow and run part 1 of the tests
+            test.asend(schd)
```

Comment:
Generators can yield data, but you can also send data into them with `generator.send`.

`asend` is the async version of `send` for async generators.

This is a sneaky trick to avoid having to write the same test twice once for reload and once for restart.